### DISCUSSION TITLE
Tracking/archiving historical sports data

### DISCUSSION THREAD

**Discussion by chisox721** [sqlalchemy/sqlalchemy#11237] [Usage Questions]
Hi all,

Was hoping to just get some general advice on how to approach a project I'm getting started on. The goal, generally speaking, is to be able to query a sports statistics database for all sorts of different bits of information regarding teams, players, games, etc. The bulk of the data is event logs pertaining to individual games. 

The area I'm struggling with is how to track changes. For example, let's say I insert a new row for a player that has been drafted into the NBA and thus a member of a team.

`Player(name='John Doe', height=78, weight=220, hometown='Dallas, Texas', team='Detroit Pistons')`

Two years later John Doe has gained 25 pounds and is now a member of the Chicago Bulls team. I'd like to be able to keep track of his weight as that does hold some significance and I most definitely need to keep track of who he's played for in the past along with who he currently plays for. I'm just kind of lost as to the best way to make that happen in a way that meshes with sqlalchemy/postgresql. For all I know I may be trying to use saw to drill a hole here but since I have some experience using both I figured it would be the best place to start at least. 

I've done a fair amount of research on how to work with historical data within the context of database management but so far haven't really been able to pin point the best approach for what I'm trying to do specifically. In my initial attempts I've resorted to just implementing phases for Team and Person models (which I'll include below for reference) but that seems off to me. I've read about versioning and sharding and while those strategies seem to be a bit outside of my comfort zone I could probably figure them out. I just don't want to get too deep into one approach and then realize I've gone down a terrible path. Any thoughts, or if someone can point me in a direction that will help me move forward, would be greatly appreciated. 

```
class Season(Base):
    __tablename__ = 'seasons'

    id: Mapped[int] = mapped_column(primary_key=True)
    year: Mapped[int]
    start_date: Mapped[datetime]
    end_date: Mapped[datetime]

    teams_phases: Mapped[List[TeamPhase]] = relationship()
    persons_phases: Mapped[List[PersonPhase]] = relationship()


class Team(Base):
    __tablename__ = 'teams'

    id: Mapped[int] = mapped_column(primary_key=True)
    slug: Mapped[str]
    name: Mapped[str]
    nickname: Mapped[str]
    fullname: Mapped[str]
    abbreviation: Mapped[str]
    color: Mapped[str]
    altcolor: Mapped[str | None]

    phases: Mapped[List[TeamPhase]] = relationship(back_populates='team')


class TeamPhase(Base):
    __tablename__ = 'team_phases'

    id: Mapped[int] = mapped_column(primary_key=True)
    team_id: Mapped[int] = mapped_column(ForeignKey('teams.id'))
    season_id: Mapped[int] = mapped_column(ForeignKey('seasons.id'))

    team: Mapped[Team] = relationship()
    season: Mapped[Season] = relationship(back_populates='teams_phases')
    persons_phases: Mapped[List[PersonPhase]] = relationship()


class Person(Base):
    __tablename__ = 'persons'

    id: Mapped[int] = mapped_column(primary_key=True)
    firstname: Mapped[str]
    lastname: Mapped[str]
    hometown: Mapped[str | None]
    birthdate: Mapped[datetime]

    phases: Mapped[List[PersonPhase]] = relationship(back_populates='person')


class PersonPhase(Base):
    __tablename__ = 'person_phases'

    id: Mapped[int] = mapped_column(primary_key=True)
    person_id: Mapped[int] = mapped_column(ForeignKey('persons.id'))
    season_id: Mapped[int] = mapped_column(ForeignKey('seasons.id'))
    team_phase_id: Mapped[int] = mapped_column(ForeignKey('team_phases.id'))
    height: Mapped[int]
    weight: Mapped[int]

    person: Mapped[Person] = relationship(back_populates='phases')
    season: Mapped[Season] = relationship(back_populates='persons_phases')
    team_phase: Mapped[TeamPhase] = relationship(back_populates='persons_phases')
```

**Comment by CaselIT** [comment]
Hi,

I think there are many ways of solving it.
You seem to already have split the "static data" from the variable ones, using the phrase tables.

I would add a `from`/`to` date columns there to track a range of times. The current one would have `to` set to null. When a new data needs to be added you can set a date in the `to` column of the current entry then add a new entry with the updated data. In postgresql you can also use a range column for this.

You may also trac the different measures (weight, team, etc) independently from each other, using either different tablets of a single measure table that has the name of the measure 

Overall there are many different ways of designing a sql schema for storing such data. You can probably find many solutions searching for database designs to store history data or time series data.
You should be able to make use of any of them with sqlalchemy
